### PR TITLE
Manage state across njs invocations with state stored in variables

### PR DESCRIPTION
**Pull request by 4141done**
### Proposed changes
Changes the mechanism by which profiling information is stored over the course of a request from in-memory in njs to strings serialized to nginx variables.

#### Why?
The njs in-memory approach works fine if you're just tracking timing and memory usage within the context of a single njs invocation.  However, if your njs code is spread across various `js_set`, `js_header_filter`, and `js_content` calls as is more often the case, that approach would fail.

#### Implementation notes
This implementation prioritizes storing profiling data as efficiently as possible. Thus profiler data is:
- Serialized in a compact format to a variable called `$profiler_internal_events`.  It looks like this: `1|profiler:start|1671143290831|47600|3||2|main_func|1671143290831|47600|3||2|js_var|1671143290831|47600|3`.  The example contains three events: one `start` and two `snapshots`.
- Never deserialized over the course of a request. We just push new events on.  Any deserialization is done at the end of the request when `collect` is called.

#### Developer Experience notes.
Now the library ships with nginx config as well as njs code.  This means we needed a pattern.  The following conventions are established in this change:

- Libraries copy to `./njs_modules/<library-name>`
- The subfolder `library-name` contains scripts.
- The subfolder `library-name` contains a `conf` folder with supporting nginx configuration files.

Check the `README.md`, but this means that the new minimal usage looks like this:

In `nginx.conf`:
```nginx
http {
  js_import script from script.mjs;

  include njs_modules/njs-memory-profiler/conf/profiler_vars.conf;

  server {
       listen       4000;
       
       location / {
        js_content script.myHandler;
       }
    }
}
```

in `script.mjs`:
```javascript
import profiler from "./njs_modules/njs-memory-profiler/njs-memory-profiler.js";

function myHandler(r) {
  profiler.init(r);
}

export default { myHandler };
```

#### Output

Output looks like this for each request:
```json
{
  // A random id generated by the profiler to tie together events for this request
  "id": "cc7da804b5a8fdd9b803a87965cde018",
  // The `events` key contains all the actual profiling events
  "events": [
    {
      // There are three types of events: `profiler:start`, `profiler:snapshot`, `profiler:end`
      "type": "profiler:start",
      
      // Arbitrary name for the event.  For start and end they will default to 
      // `profiler:start` and `profiler:end`
      "name": "profiler:start",

      // Unix timestamp for the event. In this case, when the profiler was initialized
      "createdAt": 1671144443913,

      // Size in bytes allocated to the njs vm at this point
      "size": 47600,

      // Number of blocks of memory allocated to the njs vm at this point
      "nblocks": 3
    },
    {
      "type": "profiler:snapshot",
      "name": "main_func",
      "createdAt": 1671144443913,
      "size": 47600,
      "nblocks": 3
    },
    {
      "type": "profiler:snapshot",
      "name": "js_var",
      "createdAt": 1671144443913,
      "size": 47600,
      "nblocks": 3
    },
    {
      "name": "profiler:end",
      "type": "profiler:end",
      "createdAt": 1671144443913,
      "size": 47600,
      "nblocks": 3
    }
  ]
}
```
### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [x] I have written my commit messages in the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) format.
- [x] I have formatted the code using prettier (`npx prettier -w .`)
- [x] I have read the [CONTRIBUTING](/CONTRIBUTING.md) doc
- [ ] I have added tests (when possible) that prove my fix is effective or that my feature works
- [x] I have checked that all tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto `main`

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -1,9 +1,11 @@
 ### Proposed changes
+
 Describe the use case and detail of the change. If this PR addresses an issue
 on GitHub, make sure to include a link to that issue here in this description
 (not in the title of the PR).
 
 ### Checklist
+
```
Comment (4141done) [inline .github/pull_request_template.md:8]: *annotation*:
Just formatting changes in this file

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -47,26 +47,27 @@ feature request issue first to start a discussion about the feature.
...
+  write a good message instead of `git commit -am`)
 
 ### Code Style Guide
-* Code should be njs-compatible Javascript without the need for any transpilation.
-* All code submitted should be formatted using [Prettier](https://prettier.io/) with the base configuration. You can do this by running `npx prettier -w .`
-* Where feasible, include unit and/or integration tests.
+
+- Code should be njs-compatible Javascript without the need for any transpilation.
+- All code submitted should be formatted using [Prettier](https://prettier.io/) with the base configuration. You can do this by running `npx prettier -w .`
+- Where feasible, include unit and/or integration tests.
```
Comment (4141done) [inline CONTRIBUTING.md:73]: *annotation*:
Just formatting changes in this file

File: package.json [hunk 3]
```diff
@@ -1,23 +1,24 @@
...
+    "keywords": [
+        "njs",
+        "nginx-njs"
+    ],
+    "author": "Javier Evans",
+    "license": "Apache 2.0",
+    "devDependencies": {
+        "prettier": "^2.8.1",
+        "release-please": "^14.17.3"
+    }
```
Comment (4141done) [inline package.json]: *annotation*
Note sure why I wound up with a diff here.  Just bumped the version to `0.1.3`

File: scripts/postinstall.js [hunk 4]
```diff
@@ -12,16 +12,53 @@ const moduleBase =
...
 }
 
-doCopy();
+async function removeExisting(existingBaseDir) {
+  try {
+    await fs.rm(existingBaseDir, { recursive: true });
+  } catch (_e) {
+    // We don't care if the folder was not there
+  }
+}
```
Comment (4141done) [inline scripts/postinstall.js:64]: *annotation*:

Changes here are to bring in the configuration files.  Also we are now a bit smarter and make sure we can successfully write the new dir before removing the old files.

File: src/index.mjs [hunk 5]
```diff
@@ -0,0 +1,314 @@
+/**
+ * @module Profiler
+ * Single module to allow collection of timing and memory statistics. In order to
+ * maintain collection accross njs directives profiling information is stored as a text
+ * string in an nginx variable.  The format is as follows:
+ * * Everything is an event (including start and end)
+ * * Events are separated by `||`
+ * * Values within an event are separeted by `|`
```
Comment (ciroque) [inline src/index.mjs]: "separated"

File: src/index.mjs [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+
+  pushEvent(r, EVENT_TYPE_TO_STRING[`${EVENT_TYPES.START}`], EVENT_TYPES.START);
+  r.variables.profiler_status = PROFILER_STATUS.RUNNING;
+
+  // We run final collection on vm exit, but you can manually
+  // invoke collection earlier by manually calling `collect`.
+  // If this is done, we don't collect again on vm exit.
+  njs.on("exit", () => {
+    // No async work in this context
+    collect(r, reporterFn);
```
Comment (ciroque) [inline src/index.mjs:147]: In the docs it says `collect` must be called manually, does this remove that requirement?

File: src/index.mjs [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+  });
+}
+
+/**
+ * Finishes the profiling and generates the report. Can be invoked anytime.  If it is
+ * not called explicitly, it will be invoked on the `exit` event emitted by the njs vm.
```
Comment (ciroque) [inline src/index.mjs:153]: OIC :-)

File: src/index.mjs [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+  reporterFn(
+    {
+      id: r.variables.profiler_request_id,
+      events: events,
+    },
+    r
+  );
+
+  // Stops other `collect` calls from initiating
+  r.variables.profiler_status = PROFILER_STATUS.DONE;
```
Comment (ciroque) [inline src/index.mjs:188]: 👍

**Review by ciroque — APPROVED** [review]
Couple of comments, one typo. LGTM over all :-)

File: src/index.mjs (hunk 6 above)
Comment (4141done) [reply src/index.mjs:147]: I see you got it below, but let me clarify that in the readme since if it confused you it will confuse others.

**Pull request merged**